devito/types/tensor.py

```diff
@@ -0,0 +1,597 @@
+from collections import OrderedDict
+
+import sympy
+import numpy as np
+from sympy.core.sympify import converter as sympify_converter
+from sympy.core.decorators import call_highest_priority
+
+from devito.finite_differences import Differentiable
+from devito.types.basic import AbstractTensor, Basic
+from devito.types.dense import Function, TimeFunction
+from devito.types.utils import NODE
+
+__all__ = ['TensorFunction', 'TensorTimeFunction', 'VectorFunction', 'VectorTimeFunction']
+
+
+class TensorFunction(AbstractTensor, Differentiable):
+    """
+    Tensor valued Function represented as a Matrix.
+    Each component is a Function or TimeFunction.
+
+    A TensorFunction and the classes that inherit from it takes the same parameters as
+    a DiscreteFunction and additionally:
+
+    Parameters
+    ----------
+    symmetric : bool, optional
+        Whether the tensor is symmetric or not. Defaults to True.
+    diagonal : Bool, optional
+        Wether the tensor is diagonal or not. Defaults to False.
+    staggered: tuple of Dimension, optional
+        Staggering of each component, needs to have the size of the tensor. Defaults
+        to the Dimensions.
+    """
+    _sub_type = Function
+    _op_priority = Differentiable._op_priority + 2.
+    _class_priority = 10
+
+    def __init_finalize__(self, *args, **kwargs):
+        self._is_symmetric = kwargs.get('symmetric', True)
+        self._is_diagonal = kwargs.get('diagonal', False)
+        self._staggered = kwargs.get('staggered', self.space_dimensions)
+        self._grid = kwargs.get('grid')
+        self._space_order = kwargs.get('space_order', 1)
+
+    @classmethod
+    def __subfunc_setup__(cls, *args, **kwargs):
+        """
+        Creates the components of the TensorFunction
+        either from input or from input Dimensions.
+        """
+        comps = kwargs.get("components")
+        if comps is not None:
+            return comps
+        funcs = []
+        grid = kwargs.get("grid")
+        if grid is None:
+            dims = kwargs.get('dimensions')
+            if dims is None:
+                raise TypeError("Need either `grid` or `dimensions`")
+        else:
+            dims = grid.dimensions
+        stagg = kwargs.get("staggered", None)
+        name = kwargs.get("name")
+        symm = kwargs.get('symmetric', True)
+        # Fill tensor, only upper diagonal if symmetric
+        for i, d in enumerate(dims):
+            funcs2 = [0 for _ in range(i)] if symm else []
```

Comment:
Don't we like to call such things `processed`?

Reply:
Each element is a Function so I like that name here, it's not a compiler pass or anything it just creates a list fo function.